solidity/test/lib/mailboxes.ts

```diff
@@ -24,40 +34,94 @@ export const dispatchMessage = async (
 };
 
 export const dispatchMessageAndReturnProof = async (
-  outbox: TestOutbox,
+  mailbox: TestMailboxV2,
   destination: number,
   recipient: string,
   messageStr: string,
-): Promise<MerkleProof> => {
-  const { leafIndex, message } = await dispatchMessage(
-    outbox,
+): Promise<MessageAndProof> => {
+  const nonce = await mailbox.count();
+  const { message, messageId } = await dispatchMessage(
+    mailbox,
     destination,
-    recipient,
+    utils.addressToBytes32(recipient),
     messageStr,
   );
-  const index = leafIndex.toNumber();
-  const messageHash = utils.messageHash(message, index);
-  const root = await outbox.root();
-  const proof = await outbox.proof();
+  const proof = await mailbox.proof();
   return {
-    root,
-    proof: proof,
-    leaf: messageHash,
-    index,
+    proof: {
+      branch: proof,
+      leaf: messageId,
+      index: nonce.toNumber(),
+    },
     message,
   };
 };
 
-export interface MerkleProof {
-  root: string;
-  proof: string[];
-  leaf: string;
-  index: number;
-  message: string;
+// Signs a checkpoint with the provided validators and returns
+// the signatures ordered by validator index
+export async function signCheckpoint(
+  root: types.HexString,
+  index: number,
+  mailbox: types.Address,
+  orderedValidators: Validator[],
+): Promise<string[]> {
+  const signedCheckpoints = await Promise.all(
+    orderedValidators.map((validator) =>
+      validator.signCheckpointV2(root, index, mailbox),
+    ),
+  );
+  return signedCheckpoints.map(
+    (signedCheckpoint) => signedCheckpoint.signature as string, // cast is safe because signCheckpoint serializes to hex
+  );
+}
+
+export async function dispatchMessageAndReturnMetadata(
+  mailbox: TestMailboxV2,
+  destination: number,
+  recipient: string,
+  messageStr: string,
+  orderedValidators: Validator[],
+): Promise<MessageAndMetadata> {
+  const proofAndMessage = await dispatchMessageAndReturnProof(
+    mailbox,
+    destination,
+    recipient,
+    messageStr,
+  );
+  const root = await mailbox.root();
+  const index = await mailbox.count();
```

Comment:
are checkpoint indices back to being 1-indexed? Iirc we made them 0-indexed in v1?

Reply:
Yeah seeing they're also supposed to be 0-indexed in v2 https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/v2/solidity/contracts/MailboxV2.sol#L187

Reply:
Good catch